ecosystem/typescript/sdk_v2/src/bcs/serializer.ts

```diff
@@ -0,0 +1,241 @@
+  MAX_U16_NUMBER,
+  MAX_U32_NUMBER,
+  MAX_U64_BIG_INT,
+  MAX_U8_NUMBER,
+  MAX_U256_BIG_INT,
+} from "./consts";
+import { Bytes, Uint16, Uint32, Uint8 } from "./types";
+import { AnyNumber } from "../types";
+
+export class Serializer {
+  private buffer: ArrayBuffer;
+
+  private offset: number;
+
+  constructor() {
+    this.buffer = new ArrayBuffer(64);
+    this.offset = 0;
+  }
+
+  private ensureBufferWillHandleSize(bytes: number) {
+    while (this.buffer.byteLength < this.offset + bytes) {
+      const newBuffer = new ArrayBuffer(this.buffer.byteLength * 2);
+      new Uint8Array(newBuffer).set(new Uint8Array(this.buffer));
+      this.buffer = newBuffer;
+    }
+  }
+
+  protected serialize(values: Bytes) {
+    this.ensureBufferWillHandleSize(values.length);
+    new Uint8Array(this.buffer, this.offset).set(values);
+    this.offset += values.length;
+  }
+
+  private serializeWithFunction(
+    fn: (byteOffset: number, value: number, littleEndian?: boolean) => void,
+    bytesLength: number,
+    value: number,
+  ) {
+    this.ensureBufferWillHandleSize(bytesLength);
+    const dv = new DataView(this.buffer, this.offset);
+    fn.apply(dv, [0, value, true]);
+    this.offset += bytesLength;
+  }
+
+  /**
+   * Serializes a string. UTF8 string is supported. Serializes the string's bytes length "l" first,
+   * and then serializes "l" bytes of the string content.
+   *
+   * BCS layout for "string": string_length | string_content. string_length is the bytes length of
+   * the string that is uleb128 encoded. string_length is a u32 integer.
+   *
+   * @example
+   * ```ts
+   * const serializer = new Serializer();
+   * serializer.serializeStr("çå∞≠¢õß∂ƒ∫");
+   * assert(serializer.getBytes() === new Uint8Array([24, 0xc3, 0xa7, 0xc3, 0xa5, 0xe2, 0x88, 0x9e,
+   * 0xe2, 0x89, 0xa0, 0xc2, 0xa2, 0xc3, 0xb5, 0xc3, 0x9f, 0xe2, 0x88, 0x82, 0xc6, 0x92, 0xe2, 0x88, 0xab]));
+   * ```
+   */
+  serializeStr(value: string): void {
+    const textEncoder = new TextEncoder();
+    this.serializeBytes(textEncoder.encode(value));
+  }
+
+  /**
+   * Serializes an array of bytes.
+   *
+   * BCS layout for "bytes": bytes_length | bytes. bytes_length is the length of the bytes array that is
+   * uleb128 encoded. bytes_length is a u32 integer.
+   */
+  serializeBytes(value: Bytes): void {
+    this.serializeU32AsUleb128(value.length);
+    this.serialize(value);
+  }
+
+  /**
+   * Serializes an array of bytes with known length. Therefore length doesn't need to be
+   * serialized to help deserialization.  When deserializing, the number of
+   * bytes to deserialize needs to be passed in.
+   */
+  serializeFixedBytes(value: Bytes): void {
+    this.serialize(value);
+  }
+
+  /**
+   * Serializes a boolean value.
+   *
+   * BCS layout for "boolean": One byte. "0x01" for True and "0x00" for False.
+   */
+  serializeBool(value: boolean): void {
+    if (typeof value !== "boolean") {
+      throw new Error("Value needs to be a boolean");
+    }
```

Comment:
Interesting that this is validated in boolean, but not in the others?

Oh, I guess there are checks in `checkNumberRange`